REST Server/Client allows configuring the removal of trailing slashs

Sgitario opened the pull request:
Relates to https://github.com/quarkusio/quarkus/issues/41689#issuecomment-2422198717

Before these changes, the REST server and client extensions were removing the trailing slashes that were set by users in the `@Path` annotations.

For example, when creating a resource like:

```java
@Path("/a/")
@Produces(MediaType.TEXT_PLAIN)
public class HelloResource {

        @GET
        public String echo() {
            // ...
        }
}
```

The effective path was `/a` instead of `/a/`. Note that it does not matter whether we place the `@Path` annotation at method level because the behaviour will be the same.

At the client side, when having the following client:

```java
@RegisterRestClient(configKey = "test")
@Path("/a/") <1>
public interface HelloClient {
        @GET
        @Produces(MediaType.TEXT_PLAIN)
        @Path("/b/")  <2>
        String echo();
}
```

In this case, the trailing slash will be removed from <1> but not from <2>.

After these changes, we are adding the following properties:
- `quarkus.resteasy-reactive.removes-trailing-slash`

To configure the server extension to not remove any trailing slash from the `@Path` annotations.

- `quarkus.rest-client.removes-trailing-slash`

To configure the client extension to not remove any traling slash from the `@Path` annotations used at interface level. The annotations set at method level will behave the same.

Note that this does not introduce any change in behaviour, so everything should keep working as before unless users use the new properties.

geoand commented:
I assume by

```
quarkus.resteasy-reactive.removes-trailing-slash
```

you mean

```
quarkus.resteasyremoves-trailing-slash
```

?

geoand commented on extensions/resteasy-reactive/rest-client-jaxrs/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/RestClientRemoveTrailingSlashBuildItem.java:
```diff
@@ -0,0 +1,9 @@
+/**
+ * This build item configures all the REST Clients to remove the trailing slash from the paths.
+ */
+public final class RestClientRemoveTrailingSlashBuildItem extends MultiBuildItem {
```
What's the usecase to force all clients to do this? 

Shouldn't this be a per-client setting?

Sgitario replied:
I tried to do the same as for disable smart default produces, which is also done for all the clients. 
But will update the PR to do this per-client instead +1

geoand replied:
🙏🏽

Sgitario commented:
> Relates to [#41689 (comment)](https://github.com/quarkusio/quarkus/issues/41689#issuecomment-2422198717)
> 
> Before these changes, the REST server and client extensions were removing the trailing slashes that were set by users in the `@Path` annotations.
> 
> For example, when creating a resource like:
> 
> ```java
> @Path("/a/")
> @Produces(MediaType.TEXT_PLAIN)
> public class HelloResource {
> 
>         @GET
>         public String echo() {
>             // ...
>         }
> }
> ```
> 
> The effective path was `/a` instead of `/a/`. Note that it does not matter whether we place the `@Path` annotation at method level because the behaviour will be the same.
> 
> At the client side, when having the following client:
> 
> ```java
> @RegisterRestClient(configKey = "test")
> @Path("/a/") <1>
> public interface HelloClient {
>         @GET
>         @Produces(MediaType.TEXT_PLAIN)
>         @Path("/b/")  <2>
>         String echo();
> }
> ```
> 
> In this case, the trailing slash will be removed from <1> but not from <2>.
> 
> After these changes, we are adding the following properties:
> 
> * `quarkus.rest.removes-trailing-slash`
> 
> To configure the server extension to not remove any trailing slash from the `@Path` annotations.
> 
> * `quarkus.rest-client.removes-trailing-slash`
> 
> To configure the client extension to not remove any traling slash from the `@Path` annotations used at interface level. The annotations set at method level will behave the same.
> 
> Note that this does not introduce any change in behaviour, so everything should keep working as before unless users use the new properties.

Sgitario commented:
> I assume by
> 
> ```
> quarkus.resteasy-reactive.removes-trailing-slash
> ```
> 
> you mean
> 
> ```
> quarkus.resteasyremoves-trailing-slash
> ```
> 
> ?

Actually, it's `quarkus.rest.removes-trailing-slash`. My changes took in just when this changed in main :). PR description updated.

Sgitario commented on independent-projects/resteasy-reactive/client/processor/src/main/java/org/jboss/resteasy/reactive/client/processor/scanning/ClientEndpointIndexer.java:
```diff
@@ -61,7 +61,7 @@ public ClientEndpointIndexer(AbstractBuilder builder, String defaultProduces, bo
 
     public MaybeRestClientInterface createClientProxy(ClassInfo classInfo,
-            String path) {
+            String path, boolean removesTrailingSlash) {
```
This logic would be easier to follow if the path is already updated with the start and trailing slashes logic. Wdyt?

geoand replied:
Sure

Sgitario commented:
@geoand PR updated by adding the new properties by REST client in the build time configuration.

geoand commented on extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsBuildTimeConfig.java:
```diff
@@ -69,6 +79,33 @@ interface RestClientBuildConfig {
+    /**
+     * Get the build config for a concrete REST Client implementation.
+     */
+    default Optional<RestClientBuildConfig> getConfigFor(ClassInfo restClientInterface, Optional<String> configKey) {
```
Ah sorry, I missed this was build time config.

radcortez replied:
Yes, we shouldn't be doing this.

Please use https://github.com/quarkusio/quarkus/blob/5c62c8299592a7378d7658ce0448a8598fc069ac/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsBuildTimeConfig.java#L74-L124

Sgitario replied:
I tried to use this instead of the provided instance of RestClientsBuildTimeConfig in the build step and as far as I see, the only difference is that the returned map from this method is extended with the clients using the qualified class name, but with the default values. For example, if I have the following REST Interface:

```java
@RegisterRestClient(configKey = "test")
    @Path("/slash/with/")
    public interface HelloClientUsingConfigKey {
        @GET
        @Produces(MediaType.TEXT_PLAIN)
        String echo();
    }
```

And configure it using the following property:

```
quarkus.rest-client.test.removes-trailing-slash=false
```

Using the provided instance of RestClientsBuildTimeConfig.clients(), I get only one entry with key "test" with the correct value of "removes-trailing-slash". And when using the map from the method you suggested, I get two entries: one for the key "test" with the correct value again, and another entry with the key "HelloClientUsingConfigKey", but with the default values. See the following snapshot that illustrates what I'm saying:

![Captura de pantalla de 2024-11-05 14-51-25](https://github.com/user-attachments/assets/169ee84b-1c2d-40bf-ae3a-2eca26e759bf)

Therefore, if the map from this method would return the map with all the REST Interfaces but correct values, I see the value because we could lookup the configuration by class name always. But if not, I don't see how this would help. Can you clarify?

radcortez replied:
The problem with not using `io.quarkus.restclient.config.RestClientsBuildTimeConfig#get` is that it will miss fallbacks / relocations. The previous approach caused several issues detailed in https://github.com/quarkusio/quarkus/issues/41313.

In Build Time, we are not using `@WithKeys` (but we could probably do it), so the map keeps a reference for any entry key found in the configuration, but it will always contain an entry key with the FQN, so that is the key that must be used to access the REST Client configuration.

I've just pushed a small commit to use it, plus a small fix in DevServices that I've found.

gsmet reviewed (commented):
Honest question: is it something that is clearly defined by the spec? I would expect things like that to be very precisely defined and not open to interpretation.

geoand commented:
I might be, not sure.

The client always has to deal with weird external systems which is why I am more than happy to allow users advanced options that will allow them to consume these systems

Sgitario committed:
REST Client build time fix for DevServices

Sgitario committed:
REST Server/Client allows configuring the removal of trailing slashs

Before these changes, the REST server and client extensions were removing the trailing slashes that were set by users in the `@Path` annotations.

For example, when creating a resource like:

```java
@Path("/a/")
@Produces(MediaType.TEXT_PLAIN)
public class HelloResource {

        @GET
        public String echo() {
            // ...
        }
}
```

The effective path was `/a` instead of `/a/`. Note that it does not matter whether we place the `@Path` annotation at method level because the behaviour will be the same.

At the client side, when having the following client:

```
@RegisterRestClient(configKey = "test")
@Path("/a/") <1>
public interface HelloClient {
        @GET
        @Produces(MediaType.TEXT_PLAIN)
        @Path("/b/")  <2>
        String echo();
}
```

In this case, the trailing slash will be removed from <1> but not from <2>.

After these changes, we are adding the following properties:
- `quarkus.rest.removes-trailing-slash`

To configure the server extension to not remove any trailing slash from the `@Path` annotations.

- `quarkus.rest-client.removes-trailing-slash`

To configure the client extension to not remove any traling slash from the `@Path` annotations used at interface level. The annotations set at method level will behave the same.

Note that this does not introduce any change in behaviour, so everything should keep working as before unless users use the new properties.

Sgitario commented:
@radcortez thanks for your commit! I isolated the changes you did for the DevServices issue that you found in a separated commit, and squash the rest of the changes in my commit (plus recovered a cosmetic change to compute the removesTrailingSlash). 
There is a failing test scenario in the windows build CI that seems related. Do you want me to take a look?

radcortez commented:
> There is a failing test scenario in the windows build CI that seems related. Do you want me to take a look?

Which one?

Sgitario commented:
> > There is a failing test scenario in the windows build CI that seems related. Do you want me to take a look?
> 
> Which one?

From the previous run: https://github.com/quarkusio/quarkus/pull/44299#issuecomment-2458448506
This test: `integration-tests/rest-client-reactive`:

```
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors  [error]: Build step io.quarkus.rest.client.reactive.deployment.RestClientReactiveProcessor\#setUpClientBuildTimeProperties threw an exception: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 0  at java.base/java.util.ArrayList.add(ArrayList.java:455)  at java.base/java.util.ArrayList.add(ArrayList.java:467)  at io.quarkus.restclient.config.AbstractRestClientConfigBuilder.configBuilder(AbstractRestClientConfigBuilder.java:66) - java.lang.RuntimeException
```

gsmet commented:
> The client always has to deal with weird external systems which is why I am more than happy to allow users advanced options that will allow them to consume these systems

But to make it work, you would just have to remove/add a `/` in your REST Client classes? Or am I missing something?

Sgitario commented:
> > The client always has to deal with weird external systems which is why I am more than happy to allow users advanced options that will allow them to consume these systems
> 
> But to make it work, you would just have to remove/add a `/` in your REST Client classes? Or am I missing something?

Let's add some context here. 
When writing the following REST resource using the resteasy reactive extension:

```java
@Path("/a/")
@Produces(MediaType.TEXT_PLAIN)
public class HelloResource {

        @GET
        public String echo() {
            // ...
        }
}
```

The server will be listening up `http://localhost:8080/a`, not `http://localhost:8080/a/`, because the trailing slash got removed. 
After these changes, we can avoid this removal by using `quarkus.rest.removes-trailing-slash=false`. 

For the REST client, using the reactive REST client extension and having the `@Path` annotation at class level:

```
@RegisterRestClient
@Path("/a/")
public interface HelloClient {
        @GET
        @Produces(MediaType.TEXT_PLAIN)
        String echo();
}
```

By default, the client will be trying to use `http://localhost:8080/a` instead of `http://localhost:8080/a/`. And the problem here is that the server might return 404 if we omit the trailing slash. 
After these changes, we can avoid this removal by using `quarkus.rest-client.xxx.removes-trailing-slash=false`. 

For sake of completeness, if we place the `@Path` annotation at method level instead:
```
@RegisterRestClient
public interface HelloClient {
        @GET
        @Produces(MediaType.TEXT_PLAIN)
        @Path("/a/")
        String echo();
}
```

Now, the traling slash is not removed, so it will be using `http://localhost:8080/a/`. So, this could be seen as a workaround, however when the REST interfaces are autogenerated (for example, using the OpenAPI generator), this is not always a solution. 

I hope this clarifies the need for these changes.

radcortez committed:
Avoid possible concurrency issues with RestClientsConfig.RestClientKeysProvider.KEYS in build time

geoand commented:
@radcortez can you re-review please?

Thanks

gsmet commented:
> The server will be listening up http://localhost:8080/a, not http://localhost:8080/a/, because the trailing slash got removed.

I'm a bit confused by this? Is this per spec? Because I could see how you would listen to both but I find it surprising that we don't respect the developer's intent?

radcortez commented:
Coding-wise, I think it is okay, but I agree with @gsmet. Is this really the way to go?

Sgitario commented:
> > The server will be listening up http://localhost:8080/a, not http://localhost:8080/a/, because the trailing slash got removed.
> 
> I'm a bit confused by this? Is this per spec? Because I could see how you would listen to both but I find it surprising that we don't respect the developer's intent?

I'm not an expert in the JAX-RS specs, but according to the [version 1.0](https://download.oracle.com/otn-pub/jcp/jaxrs-1.0-fr-eval-oth-JSpec/jaxrs-1.0-final-spec.pdf) and [version 2.0](https://download.oracle.com/otn-pub/jcp/jaxrs-2_0-fr-eval-spec/jsr339-jaxrs-2.0-final-spec.pdf) where it says:

```
If the resulting string ends with ‘/’ then remove the final character
```

This is in the section "Converting URI Templates to Regular Expressions".
So, I would say this is an expected behaviour and matches what the spec says.

geoand commented:
I am not sure what the deal with failing test is...

Sgitario commented:
> I am not sure what the deal with failing test is...

@geoand how do you feel about merging these changes? I don't think the test failures are related.

geoand commented:
Right, CI is green now so let's merge
